host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,235 @@
+
+            result = ENCLU_ERESUME;
+            break;
+        }
+    }
+
+done:
+    /* If the result <= 0, the value will be forwared as the return
+     * value of _vdso_sgx_enter_enclave. Otherwise, _vdso_sgx_enter_enclave
+     * will invoke the ENCLU[result] instead of returning to the caller. */
+    return result;
+}
+
+/* The function should never be inline to preserve the stack frame. */
+OE_NEVER_INLINE
+oe_result_t oe_vdso_enter(
+    void* tcs,
+    uint64_t arg1,
+    uint64_t arg2,
+    uint64_t* arg3,
+    uint64_t* arg4,
+    oe_enclave_t* enclave)
+{
+    oe_ecall_context_t ecall_context = {0};
+    oe_result_t result = OE_UNEXPECTED;
+    struct sgx_enclave_run run = {0};
+    vdso_args_t return_args = {0};
+    int return_value = 0;
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+
+    oe_setup_ecall_context(&ecall_context);
+
+    run.tcs = (uint64_t)tcs;
+    run.user_handler = (uint64_t)oe_vdso_user_handler;
+    run.user_data = (uint64_t)&return_args;
+
+    while (1)
+    {
+        /* Compiler will usually handle this on exiting a function that uses
+         * AVX, but we need to avoid the AVX-SSE transition penalty here
+         * manually as part of the transition to enclave. See
+         * https://software.intel.com/content/www/us/en/develop/articles
+         * /avoiding-avx-sse-transition-penalties.html */
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        // Define register bindings and initialize the registers.
+        OE_DEFINE_REGISTER(rax, _vdso_sgx_enter_enclave);
+        OE_DEFINE_REGISTER(rbx, &run);
+        OE_DEFINE_REGISTER(rcx, ENCLU_EENTER);
+        OE_DEFINE_REGISTER(rdx, &ecall_context);
+        OE_DEFINE_REGISTER(rdi, arg1);
+        OE_DEFINE_REGISTER(rsi, arg2);
+
+        /* Save and restore MXCSR, x87 control word, RFLAGS, and non-volatile
+         * registers (excpet for RSP and RBP that are expected to be preserved)
+         * before and after the vDSO call conform the Linux x64 ABI */
+        asm volatile("stmxcsr %[mxcsr] \n\t" // Save MXCSR
+                     "fstcw %[fcw] \n\t"     // Save x87 control word
+                     "pushfq \n\t"           // Save RFLAGS
+                     "pushq %%rbx \n\t" // Pass 7th argument (run) via stack
+                     "call *%%rax \n\t"
+                     "popq %%rbx \n\t" // Restore the stack
+                     "popfq \n\t"      // Restore RFLAGS
+                     "mov %%rax, %[return_value] \n\t"
+                     "fldcw %[fcw] \n\t"     // Restore x87 control word
+                     "ldmxcsr %[mxcsr] \n\t" // Restore MXCSR
+                     : OE_VDSO_REGISTERS, [return_value] "=m"(return_value)
+                     : [fcw] "m"(fcw), [mxcsr] "m"(mxcsr)
+                     : OE_VDSO_CLOBBERED_REGISTERS);
+
+        if (return_value < 0)
+            OE_RAISE(OE_FAILURE);
+
+        /* Update arg1 and arg2 with outputs returned by the enclave */
+        arg1 = return_args.rdi;
```

Comment:
I think this is fine because the only flow goes to here is when the `oe_vdso_user_handler` returns 0, which is the case of ENCLU_EEXIT:
```
            return_args->rdi = arg1;
            return_args->rsi = arg2;
            result = 0;
```

In this case, the return_args will always be override with latest return values from the enclave.